rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,197 @@
+    m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+) -> c_types::c_int {
+    // SAFETY: `v` was created by a previous call to `next_callback` or
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `Box<Peekable<T::Iterator>>::into_raw`.
+    let iterator = unsafe { (v as *mut Peekable<T::Iterator>).as_mut() };
+    if let Some(iterator) = iterator {
+        let s = match iterator.peek() {
+            // TODO: Replace with fallible `to_string` when available.
+            Some(item) => item.to_string() + "\0",
+            None => "\0".to_string(),
+        };
+        // SAFETY: Calling a C function. `s` is guaranteed to be null terminated
+        // because we explicitly constructed it just above.
+        unsafe {
+            bindings::seq_puts(m, s.as_ptr() as *const u8 as *const c_types::c_char);
+        }
+    }
+    0
+}
+
+extern "C" fn start_callback<T: SeqOperations>(
+    m: *mut bindings::seq_file,
+    pos: *mut bindings::loff_t,
+) -> *mut c_types::c_void {
+    // SAFETY: This function will be called by opening a proc file generated
+    // by a call to `proc_create::<T>`, so `m` can be converted using
+    // `SeqFile<T>`.
+    let arg = unsafe { &*SeqFile::<T>::convert(m) };
+    // SAFETY: The caller guarantees that `pos` points to a valid `loff_t`.
+    let pos = unsafe { *pos };
+    match T::start(arg) {
+        Some(mut wrapper) => {
+            for _ in 0..pos {
+                wrapper.as_mut().next();
+            }
+            Box::into_raw(wrapper) as *mut c_types::c_void
+        }
+        None => ptr::null_mut(),
+    }
+}
+
+struct SeqFileOperationsVTable<T>(PhantomData<T>);
+
+impl<T: SeqOperations> SeqFileOperationsVTable<T> {
+    const VTABLE: bindings::seq_operations = bindings::seq_operations {
+        start: Some(start_callback::<T>),
+        stop: Some(stop_callback::<T>),
+        next: Some(next_callback::<T>),
+        show: Some(show_callback::<T>),
+    };
+}
+
+/// A `seq_file` referencing data of type `T`.
+pub struct SeqFile<T> {
+    seq_ops: bindings::seq_operations,
+    _pin: PhantomPinned,
+    context: T,
+}
+
+impl<T> SeqFile<T> {
+    fn new_pinned(seq_ops: bindings::seq_operations, context: T) -> KernelResult<Pin<Box<Self>>> {
+        Ok(Pin::from(Box::try_new(SeqFile {
+            seq_ops,
+            _pin: PhantomPinned,
+            context,
+        })?))
+    }
+
+    /// Retrieve the [`SeqFile::context`] associated with a [`bindings::seq_file`].
+    ///
+    /// # Safety
+    ///
+    /// `m` must have been created from a proc file generated by calling
+    /// `proc_create::<T>`.
+    unsafe fn convert(m: *mut bindings::seq_file) -> *const T {
+        let reg = crate::container_of!((*m).op, Self, seq_ops);
+        &(*reg).context
+    }
+}
+
+/// Create a `/proc` file.
+///
+/// The returned value must not be dropped until the module is unloaded.
+pub fn proc_create<T: SeqOperations + Sync>(
```

Comment:
Do we need `#[cfg]` on `proc` etc.?

Reply:
(Not necessarily on the function itself, it could be on the implementation instead so that callers do not need to care, i.e. skipping the bindings call if `PROC` is not there -- please check what the kernel does currently for `PROC`-related calls).

Reply:
Hmmm... It looks like other places have an `#ifdef CONFIG_PROC_FS` block around all the code for setting up the `seq_file`. Not sure which way makes more sense?

Reply:
Ended up splitting this into `seq_file.rs` and `proc_fs.rs`. The proc specific parts are all under `#[cfg(CONFIG_PROC_FS)]`, but the `seq_file` parts aren't so we can later add e.g. debugfs `seq_file`s.